### ISSUE TITLE
explore a way to add deferred datasets to tools

### ISSUE THREAD

**Issue by volodymyrss** [esg-epfl-apc/contributed-galaxy-workflows#32]
@francoismg  could you refer here to the context and explain the limitations?

https://github.com/esg-epfl-apc/astro-archive

**Comment by bgruening** [comment]
You mean add deferred datasets to tools?

**Comment by volodymyrss** [comment]
Yes!

**Comment by bgruening** [comment]
You can't. Tools are independent of where the data is or how the data is referenced. This is part of the framework and later the job, that might be responsible for staging the data before the tool runs. So it's a framework decision. You need to specify it during upload into Galaxy.

**Comment by volodymyrss** [comment]
But what if the upload to galaxy is done by the tool?

**Comment by volodymyrss** [comment]
But what if the upload to galaxy is done by the tool?

**Comment by bgruening** [comment]
I don't think we have a solution for that until now. Is the result of your tool always a file (stable URL to a file)?

@jmchilton to you have any idea here?

In practice maybe that is not a big problem. The tool will download the data and store it into a "nearby" S3 - at least that is the ESG use-case. And the n+1 tool gets it from this S3 and puts the n+1-results into the S3 again, without going to Freiburg.

**Comment by francoismg** [comment]
> I don't think we have a solution for that until now. Is the result of your tool always a file (stable URL to a file)?
> 
> @jmchilton to you have any idea here?
> 
> In practice maybe that is not a big problem. The tool will download the data and store it into a "nearby" S3 - at least that is the ESG use-case. And the n+1 tool gets it from this S3 and puts the n+1-results into the S3 again, without going to Freiburg.

There's is the upload.py way but you said you wouldn't recommend that right? After that you mentioned using the galaxy api so we could make a visualization that make a call to the "tool/upload_file" route with the deferred properties set to true, is that correct?

**Comment by bgruening** [comment]
Yes that is correct. If you use the upload API you can set the deferred dataset flag.

**Comment by francoismg** [comment]
> Yes that is correct. If you use the upload API you can set the deferred dataset flag.

ok thanks, and when you make an api call from a visualization you don't need and api key right?

**Comment by bgruening** [comment]
You don't need to, you are already inside the user-session.

**Comment by volodymyrss** [comment]
> I don't think we have a solution for that until now. Is the result of your tool always a file (stable URL to a file)?

Yes, it is a stable URL to a file. Well, it's conditional in inputs, for certain inputs it is always so.

> @jmchilton to you have any idea here?
> 
> In practice maybe that is not a big problem. The tool will download the data and store it into a "nearby" S3 - at least that is the ESG use-case. And the n+1 tool gets it from this S3 and puts the n+1-results into the S3 again, without going to Freiburg.

It would be faster if the data is not  downloaded until the n+1 tool actually runs.  If n+1 tool runs, for example, on ARC, ARC can download the data itself.

Can rucio approach taken by OakRidge group also use this approach instead of the workaround they used?

Also I suppose we want a sustainable solution, so supporting own storage.

Could you clarify what is the limitation of ESG use-case?

**Comment by volodymyrss** [comment]
> Yes that is correct. If you use the upload API you can set the deferred dataset flag.

The visualization upload is a different case, it is useful too, but we still need data output by tools to be deferred sometimes, since doing selection in the tool is the way to make it reproducible (who knows what user clicked in the UI, and we need to be able to track provenance)

By the way, is there some way to make a "contract" on URL stability? Guaranteed-by, for so-long. Stability may range from, say, weeks, to decades (in the case of some DOIs). 

I do, however, appreciate that the decisions on if the data is treated as deferred, might not need to be made by the tool. 
It makes sense if they are done by the framework.

But, we know that in a reasonable situation data output by some tools and "upload" activities in UI should be deferred, because they are large but the URLs are very stable.
Is there a way for the platform to reasonably make these decisions based on some information produced by the tool/UI?

**Comment by francoismg** [comment]
> The visualization upload is a different case, it is useful too, but we still need data output by tools to be deferred sometimes, since doing selection in the tool is the way to make it reproducible (who knows what user clicked in the UI, and we need to be able to track provenance)

Regarding reproducibility in visualization plugin, we could have an option that let the user download a json file with all the parameters he chose. It could be the same as the json inputs file we generate from the tool xml so you could even pass it to the archive python script from the command line (for automation purpose or things like that I guess) and get the same results wether it's from a visualization or a standard tool (except for the deferred thing apparently). We could also add the possibility in both the visualization and tool interface to have json file input so you could just use your json parameter file and let either the visualization or the tool take care of the rest. If you think it makes sense we could add that feature to the tool.
 
> I do, however, appreciate that the decisions on if the data is treated as deferred, might not need to be made by the tool. It makes sense if they are done by the framework.

I guess ultimately it should be the user choice whether he wants some dataset to be deferred or not, if i'm not mistaken once a dataset is tagged as deferred you cannot choose to download it to turn it into a normal dataset which keeps you from using visualizations for instance. So it seems like the decision to make a dataset deferred relies a lot on what the user wants to do with it afterward I guess

**Comment by volodymyrss** [comment]
> > The visualization upload is a different case, it is useful too, but we still need data output by tools to be deferred sometimes, since doing selection in the tool is the way to make it reproducible (who knows what user clicked in the UI, and we need to be able to track provenance)
> 
> Regarding reproducibility in visualization plugin, we could have an option that let the user download a json file with all the parameters he chose. It could be the same as the json inputs file we generate from the tool xml so you could even pass it to the archive python script from the command line (for automation purpose or things like that I guess) and get the same results wether it's from a visualization or a standard tool (except for the deferred thing apparently). We could also add the possibility in both the visualization and tool interface to have json file input so you could just use your json parameter file and let either the visualization or the tool take care of the rest. If you think it makes sense we could add that feature to the tool.

That would be a lot of manual inputs.json upload/download.
Reproducibility here would work well if once the selection is made, history would represent tool execution as if it the tool was run normally.
So that when the tool is run next time, or if the history is exported as a workflow and they re-run, the same result would appear.
I guess this might be possible to do through API, to create a new history entry, not sure.

This is interesting but a bit tangential to deferred datasets issue.
What we need to learn here is if we agree that the tool should be able to output deferred datasets. It seems so. 

> 
> > I do, however, appreciate that the decisions on if the data is treated as deferred, might not need to be made by the tool. It makes sense if they are done by the framework.
> 
> I guess ultimately it should be the user choice whether he wants some dataset to be deferred or not, if i'm not mistaken once a dataset is tagged as deferred you cannot choose to download it to turn it into a normal dataset which keeps you from using visualizations for instance. So it seems like the decision to make a dataset deferred relies a lot on what the user wants to do with it afterward I guess

Once dataset is published by a user, other users may want to do other things with it.
I guess it is always possible to make a tool that downloads the dataset. In fact, this sounds generic, there is not such a thing?
It is a bit weird that it is not possible to easily get to visualization of a deferred dataset. @bgruening is saying above that tools do not care about deferred status. But UI does in such a key way? Is there a galaxy UI to "stage" the deferred dataset? 
Dataset might end up stored locally (in a nearby S3), cached by a site. So the user anyway does not fully control all copies of the datasets.
The exact nature of data movement within a platform can not be responsibility of the user.

I do not yet understand sufficiently what deferred datasets are meant to be in Galaxy and what are the crucial details of their implementation.

**Comment by volodymyrss** [comment]
> I have not touched deferred data in a while... but I think it would be a good idea if tools could produced deferred datasets. I can't see why we wouldn't want that currently. It feels like it wouldn't be a huge implementation hurdle on our end either.

Would it mean that the tool produces stable URL? Stability would be understood in some particular sense?

Or even tool would upload the data to a location elsewhere and return the URL? This last option may only make sense as a workaround when galaxy for some reason does not have a way to store the data is the necessary storage.

**Comment by francoismg** [comment]
> I have not touched deferred data in a while... but I think it would be a good idea if tools could produced deferred datasets. I can't see why we wouldn't want that currently. It feels like it wouldn't be a huge implementation hurdle on our end either.

That would be great.

Would it be directly dealt with in the outputs section with like a new "deferred = true" parameter or a specific tag like "<deferred_data />"? Would it work with "<discover_datasets>" too?

Or will we have a script like data_source.py which we would pass urls to?
